### PR TITLE
remember nav bar visibility using localstorage

### PR DESCRIPTION
**Pull request by awalGarg**
Initially discussed in https://github.com/mozilla/serviceworker-cookbook/issues/108#issuecomment-158184177, https://github.com/mozilla/serviceworker-cookbook/issues/108#issuecomment-158192392 and https://github.com/mozilla/serviceworker-cookbook/issues/108#issuecomment-158318447

### CODE REVIEW

File: src/js/layout.js [hunk 1]
```diff
@@ -28,6 +30,8 @@ function resizeIframe(iframe) {
 
 window.addEventListener('load', function() {
   document.body.classList.add('loaded');
+  var navStatus = Number(localStorage.getItem('withNav'));
+  document.querySelector('.book').classList.toggle('with-nav', navStatus);
```
Comment (darkwing) [inline src/js/layout.js]: We probably don't need `navStatus` as a var, or we could hoist it.

**Comment by darkwing** [comment]
localStorage can be a pain in private/incognito windows -- should that be a concern?

**Comment by awalGarg** [comment]
> localStorage can be a pain in private/incognito windows -- should that be a concern?

@darkwing no, that only happens if we do some convoluted operations on data returned by `.getItem` which wasn't previously set (by, say `.setItem`). Calling `Number` on `null` is perfectly valid and would simply fail to toggle the class.

**Comment by awalGarg** [comment]
r = review? Sorry not very familiar with OSS contribution ^^

**Comment by darkwing** [comment]
I like the idea, @awalGarg, but I think we should find a way to make the execution better.  The frequent slide on each link click is a bit jarring.  It may behoove us to make those top link clicks AJAX-powered so that the toggle doesn't happen.  Or we could add a session var to simply not show the sidebar on initial page load.  I'll continue to think on this.

**Comment by awalGarg** [comment]
@darkwing I agree. Moving out the toggling from window.onload to document.onDOMContentLoaded seems to fix this, like so:

```
diff --git a/src/js/layout.js b/src/js/layout.js
index c19f912..372ff41 100644
--- a/src/js/layout.js
+++ b/src/js/layout.js
@@ -28,12 +28,19 @@ function resizeIframe(iframe) {
 
 })();
 
-window.addEventListener('load', function() {
-  document.body.classList.add('loaded');
-  document.querySelector('.book').classList.toggle(
+document.addEventListener('DOMContentLoaded', function() {
+  var book = document.querySelector('.book');
+  var nav = document.querySelector('nav');
+  nav.style.display = 'hidden';
+  book.classList.toggle(
     'with-nav',
     Number(localStorage.getItem('withNav'))
   );
+  nav.style.display = 'initial';
+});
+
+window.addEventListener('load', function() {
+  document.body.classList.add('loaded');
 });
 
 (function()  {
```

**Comment by awalGarg** [comment]
err, ignore the .hidden stuff. I messed up some stuff.

**Comment by awalGarg** [comment]
@darkwing I added the cleaned up domcontentloaded patch and rebased on latest master. Can you test now? The "jarring" isn't present anymore locally.

**Comment by darkwing** [comment]
Wow @awalGarg , this is very close.  The only issue I spot is that the first viewing of the site has the sidebar hidden; we should really be showing the sidebar by default if the user hasn't been on the site yet.

**Comment by awalGarg** [comment]
Hmm, that should be an easy fix but might make the code verbose. I'll see if there is a better way in a few hours.

**Comment by darkwing** [comment]
What if we reverse the philosophy?  If we call it "hideNav" and set it to "1" to hide the nav, then the nav will display by default when the user hits the site the first time! :)

**Comment by awalGarg** [comment]
@darkwing yeah that works indeed, good call :)

**Comment by awalGarg** [comment]
Also just remembered that `.toggle` returns state anyways, so we don't need another call to `.contains`. Fixed that.

**Comment by darkwing** [comment]
Awesome, this is much better!  Spending a tiny bit more time thinking on this; the "close" is un-noticeable in Chrome but a tad slower and noticeable in Firefox.  Going to see if there's anything more I can think of.  Great work @awalGarg , I'll have an update for you on this soon!

**Comment by awalGarg** [comment]
@darkwing Hmm, I actually didn't test this in Firefox. My bad.

One way I see to solve this, is rather peculiar, and rather hacky. We know that script parsing is 100% blocking and the browser doesn't parse the page further until it completely executes current script (specially inline ones).

So, what we can do here, is move the `with-nav` class to the body from the `div.book` element. And then just before the `div.book` declaration, we can add the code which is responsible for toggling the class. Example patch:

```
diff --git a/src/css/style.css b/src/css/style.css
index f39035e..df3a07a 100644
--- a/src/css/style.css
+++ b/src/css/style.css
@@ -103,7 +103,7 @@ body {
 }
 
 @media (min-width: 600px) {
-  .book.with-nav main {
+  body.with-nav main {
     left: 300px;
   }
   .book main {
@@ -120,7 +120,7 @@ body {
     left: -100%;
     transition: left 250ms ease;
   }
-  .book.with-nav main {
+  body.with-nav main {
     transform: translate(calc(100% - 60px),0px);
   }
   .book main {
@@ -128,7 +128,7 @@ body {
   }
 }
 
-.book.with-nav nav {
+body.with-nav nav {
   left: 0;
 }
 
diff --git a/src/js/layout.js b/src/js/layout.js
index 2745003..f535e19 100644
--- a/src/js/layout.js
+++ b/src/js/layout.js
@@ -7,7 +7,7 @@ function resizeIframe(iframe) {
 (function() {
 
   document.getElementById('navToggle').addEventListener('click', function() {
-    document.querySelector('.book').classList.toggle('with-nav');
+    localStorage.setItem('hideNav', 1-Number(document.body.classList.toggle('with-nav')));
   });
 
   // This is a hacky way to highlight the current active tab. This should
diff --git a/src/tpl/layout.html b/src/tpl/layout.html
index bcb57d3..9f6cc15 100644
--- a/src/tpl/layout.html
+++ b/src/tpl/layout.html
@@ -9,8 +9,13 @@
   <link rel="stylesheet" href="bundle.css">
   <script type="text/javascript" src="bundle.js" defer></script>
 </head>
-<body>
-
+<body class="with-nav">
+<script type="text/javascript">
+  document.body.classList.toggle(
+    'with-nav',
+    1-Number(localStorage.getItem('hideNav'))
+  );
+</script>
   <script>
   (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
   (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
@@ -21,7 +26,7 @@
   ga('send', 'pageview');
   </script>
 
-  <div class="book with-nav">
+  <div class="book">
     <nav><div class="recipes-cover">
       <div id="tabzilla">
         <a href="https://www.mozilla.org/">Mozilla</a>
```

I am not sure if this is the best solution, though.

**Comment by awalGarg** [comment]
well, this patch is against my local copy of origin's master. dunno what commit it is at :/

**Comment by darkwing** [comment]
We can give this a go and monitor going forward.

**Pull request merged**